### PR TITLE
[Monaco] Add `monaco.editor.registerEditorOpener` method to be able to intercept editor open operations

### LINKED ISSUE #2000: Goto definition (Ctrl + click) works only if definition is in same model
**monaco-editor version:** 0.20.0
**Browser:** Chrome
**OS:** Windows
**Playground code that reproduces the issue:**

**Reproduction steps:**

- Go to playground
- Paste code below
- Run
- ctrl + MouseOver on IdentifierA in line 1
=> Identifier A is blue and underlined; small widget pops up showing reference "another IdentifierA" => **OK!**
- ctrl left click on IdentifierA in line 1
=> IdentifierA in line 3 blinks and Cursor jumps to line 3 => **OK!**

- ctrl + MouseOver on IdentifierB in line 2
=> Identifier B is underlined; small widget pops up showing reference "another IdentifierB" => **OK!**
- ctrl left click on IdentifierB in line 2
=> Nothing happens. Editor should display model2 and Curser should jump to line 1 => **Bug!**

**I was able to fix this bug:**
a) insert 
`editor.setModel(model);`
in standaloneCodeServiceImpl.ts after line 43
b) replace 
`return null;`
whith
`return monaco.editor.getModels().find(model => model.uri.toString() === resource.toString());`
in Method `StandaloneCodeEditorServiceImpl.findModel`

Instead of b) it would certainly be better to use `SimpleEditorModelResolverService`, but i don't know how to get a reference of it so i better refrain from making a pull request on base of my amateur work.

**Editor Playground html:**
```
<div id="container" style="height:100%;"></div>
```

**Editor Playground css:**
none

**Editor Playground js:**
```
const editor = monaco.editor.create(document.getElementById("container"), {});

const model1 = monaco.editor.createModel("identifierA\nIdentifierb\nanother IdentifierA", "python", monaco.Uri.from({
    scheme: "inmemory",
    path: "file1.macc",
}))

const model2 = monaco.editor.createModel("another Identifierb", "python", monaco.Uri.from({
    scheme: "inmemory",
    path: "file2.macc",
}))

editor.setModel(model1)

monaco.languages.registerDefinitionProvider("python", {
    provideDefinition: (model, position, token) => {
        if (position.lineNumber == 1) {
            return {
                uri: model1.uri,
                range: {
                    startLineNumber: 3,
                    startColumn: 9,
                    endLineNumber: 3,
                    endColumn: 20
                }
            }
        }

        if (position.lineNumber == 2) {
            return {
                uri: model2.uri,
                range: {
                    startLineNumber: 1,
                    startColumn: 9,
                    endLineNumber: 1,
                    endColumn: 20
                }
            }
        }
    }
});
```

### PR DESCRIPTION
**Pull request by spahnke**
Fixes https://github.com/microsoft/monaco-editor/issues/2000

By default the "go to defintion" action in the Monaco editor does nothing if the definition is in another model, or opens a peek window when multiple definitions are found. This default is a good thing because just changing the underlying model would be very detrimental when embedding the editor. However, there are use-cases where you maybe want to control what happens when you go to a definition. For example, when we embed the Monaco editor we would like to open a new tab to the source code containing the definition for certain models but not all of them.

This PR adds a new function `monaco.editor.registerEditorOpener` that lets the embedder intercept such open operations and handle those themselves. By using a boolean return value the embedder can indicate if the resource was handled or the editor should fallback to the default. The function returns an `IDisposable` instance that can be used to unregister the handler again. This interception can be tested in the code below by e.g. Ctrl-clicking on the `alert` function. By intercepting the open operation you can also choose to change the underlying editor model yourself if that makes sense in your application (see linked issue).

**Edit:** I decided to remove a second commit because it didn't work properly in some cases. This reduces the PR to just the new method to intercept the editor open operation.

**Monaco Playground Code:**
```js
const text = `function hello() {
	alert('Hello world!');
	Date.now();
}`;

// Hover on each property to see its docs!
const editor = monaco.editor.create(document.getElementById("container"), {
	value: text,
	language: "javascript",
	automaticLayout: true,
});

monaco.editor.registerEditorOpener({
	openCodeEditor(source, resource, selectionOrPosition) {
		console.log(source, resource, selectionOrPosition);
		if (resource.path === "/lib.dom.d.ts") {
			// simulate openening a new browser tab for our own type (open definition of alert)
			console.log(`Opened definition for ${resource}`);

			// alternatively set model directly in the editor if you have your own tab/navigation implementation
			// const model = monaco.editor.getModel(resource);
			// editor.setModel(model);
			// if (monaco.Range.isIRange(selectionOrPosition)) {
			// 	editor.revealRangeInCenterIfOutsideViewport(selectionOrPosition);
			// 	editor.setSelection(selectionOrPosition);
			// } else {
			// 	editor.revealPositionInCenterIfOutsideViewport(selectionOrPosition);
			// 	editor.setPosition(selectionOrPosition);
			// }

			return true;
		}
		return false;
	}
});
```

### CODE REVIEW

**Comment by WNemencha** [comment]
Thanks for the PR, i really need this :)

**Comment by spahnke** [comment]
@hediet Could this be considered for one of the next Monaco versions? Thank you!

File: src/vs/editor/standalone/browser/standaloneEditor.ts [hunk 1]
```diff
@@ -456,6 +459,36 @@ export function registerLinkOpener(opener: ILinkOpener): IDisposable {
 	});
 }
 
+export interface ICodeEditorOpener {
+	openCodeEditor(source: ICodeEditor, resource: URI, selectionOrPosition?: IRange | IPosition): boolean | Promise<boolean>;
```
Comment (hediet) [inline src/vs/editor/standalone/browser/standaloneEditor.ts:463]: Please add some doc comments explaining when this method is called (copying the comment from the register call).

**Review by hediet — DISMISSED** [review]

**Comment by spahnke** [comment]
Addressed review and rebased onto main.

**Comment by hediet** [comment]
Thanks!

> Addressed review and rebased onto main.

Generally, please don't rebase in PRs, as that resets the review.
(we can also just squash the PR when merging)

**Review by hediet — APPROVED** [review]

**Review by aiday-mar — APPROVED** [review]

**Comment by spahnke** [comment]
This seems stuck, do I need to do something?

**Review by alexr00 — APPROVED** [review]

**Pull request merged**